### DISCUSSION TITLE
Contiguous/non-contiguous decorator

### DISCUSSION THREAD

**Discussion by EmilyBourne** [pyccel/pyccel#1608] [Ideas]
@dominikbell noticed that the compilation time for Pyccel became significantly slower when we released version 1.8.0.
One reason for this is that the wrapper is now more complicated as it also handles slices. This is not useful for every application. 

Additionally knowing that an array is contiguous allows some compilers to provide additional optimisations (Fortran provides a keyword for this).

I wonder if it would be useful to add an `@contiguous` decorator? This would act similarly to the `@allow_negative_indexes` decorator and would lead to the mentioned array being marked as `contiguous` in Fortran. Additionally it would remove the need for the step parameter in the wrapper. In the example provided by @dominikbell this change reduces the time spent compiling the wrapper by half.

**Comment by yguclu** [comment]
Can you please share a minimal working example?

**Reply by EmilyBourne** [reply]
This is the example provided by Dominik:
```python
def fun1(a: 'float[:, :]', b: 'float[:]'):
    n_entries = shape(a)[0]

    for i in range(n_entries):
        a[i, 0] += 1
        b[i] = sum(a[i, :])
        a[i, 0] *= 2

    b[:] += 2

    n_entries_b = shape(b)[0]

    for i in range(n_entries_b):
        b[i] -= 2

    for i in range(n_entries):
        a[i, 0] -= 1
        b[i] *= sum(a[i, :])
```

I suggest that it becomes:
```python
@contiguous('a', 'b')
def fun1(a: 'float[:, :]', b: 'float[:]'):
    n_entries = shape(a)[0]

    for i in range(n_entries):
        a[i, 0] += 1
        b[i] = sum(a[i, :])
        a[i, 0] *= 2

    b[:] += 2

    n_entries_b = shape(b)[0]

    for i in range(n_entries_b):
        b[i] -= 2

    for i in range(n_entries):
        a[i, 0] -= 1
        b[i] *= sum(a[i, :])
```

**Reply by EmilyBourne** [reply]
Currently the `bind_c` file looks like this:
```fortran
  subroutine bind_c_fun1(bound_a, bound_a_shape_1, bound_a_shape_2, &
        bound_a_stride_1, bound_a_stride_2, bound_b, bound_b_shape_1, &
        bound_b_stride_1) bind(c)

    implicit none

    type(c_ptr), value :: bound_a
    integer(i64), value :: bound_a_shape_1
    integer(i64), value :: bound_a_shape_2
    integer(i64), value :: bound_a_stride_1
    integer(i64), value :: bound_a_stride_2
    type(c_ptr), value :: bound_b
    integer(i64), value :: bound_b_shape_1
    integer(i64), value :: bound_b_stride_1
    real(f64), pointer :: a(:,:)
    real(f64), pointer :: b(:)

    call C_F_Pointer(bound_a, a, [bound_a_shape_2 * bound_a_stride_2, &
          bound_a_shape_1 * bound_a_stride_1])
    call C_F_Pointer(bound_b, b, [bound_b_shape_1 * bound_b_stride_1])
    call fun1(a = a(1_i64::bound_a_stride_2, 1_i64::bound_a_stride_1), b &
          = b(1_i64::bound_b_stride_1))

  end subroutine bind_c_fun1
```
If the arrays were marked as contiguous this would become:
```fortran
  subroutine bind_c_fun1(bound_a, bound_a_shape_1, bound_a_shape_2, &
        bound_b, bound_b_shape_1) bind(c)

    implicit none

    type(c_ptr), value :: bound_a
    integer(i64), value :: bound_a_shape_1
    integer(i64), value :: bound_a_shape_2
    type(c_ptr), value :: bound_b
    integer(i64), value :: bound_b_shape_1
    real(f64), pointer :: a(:,:)
    real(f64), pointer :: b(:)

    call C_F_Pointer(bound_a, a, [bound_a_shape_2 , bound_a_shape_1])
    call C_F_Pointer(bound_b, b, [bound_b_shape_1])
    call fun1(a = a, b = b)

  end subroutine bind_c_fun1
```

**Reply by yguclu** [reply]
If this information is for the function arguments, I would prefer to include it in the type annotation as we currently do with `const`.

**Reply by yguclu** [reply]
By the way, I see a potential runtime benefit to specifying the memory contiguous-ness. Have you observed anything like that?

**Reply by EmilyBourne** [reply]
I have never observed this directly, but it is definitely supposed to be able to bring a benefit. That is why I suggest a decorator as this allows us to tag any array as contiguous, not just the arguments. But we could auto-deduce this property for local arrays

**Reply by said-hadjout** [reply]
I think we should instead add the concept of a strided numpy array, something like `int[::,::]` which will assume strides

**Reply by said-hadjout** [reply]
you can even specify which dimension is strided `int[:,::]`

**Reply by EmilyBourne** [reply]
You mean make non-strided the default? (Non-contiguous or strided arrays have been supported in Pyccel since version 1.8.0)

**Reply by said-hadjout** [reply]
yes, that is better in my opinion, the default is a contiguous array, if the user wants a strided array, he needs to specify that, using the syntax mentioned before, where he has more flexibility in choosing a specific dimension to be strided.
In my usage of pyccel, I have never felt the need to use a strided array. So we are good as long as there are clear error messages when the user passes the wrong array type.

**Reply by said-hadjout** [reply]
> By the way, I see a potential runtime benefit to specifying the memory contiguous-ness. Have you observed anything like that?

For the runtime benefit, I think we can get some performance because we are doing the memory allocation outside of Fortran using Python, but I don't have any example

**Reply by EmilyBourne** [reply]
> I think we can get some performance because we are doing the memory allocation outside of Fortran using Python

This is true for both the contiguous and the non-contiguous case

**Reply by saidctb** [reply]
yes, but in the contiguous case, we are telling the compiler that the memory is contiguous, otherwise it will not know, when the array is allocated in Python and it will assume the general case

**Reply by EmilyBourne** [reply]
This is something I need to test at some point. Does assuming the general case actually lead to slower performance?

**Reply by saidctb** [reply]
probably it will not allow some optimizations, you have a discussion here https://fortran-lang.discourse.group/t/when-is-contiguous-useful/3082

**Comment by saidctb** [comment]
If you add the `contiguous` keyword in a function argument and pass a non-contiguous array, Fortran will create a copy of the array
